Code/Framework/AzCore/AzCore/Asset/AssetManager.cpp

```diff
@@ -567,7 +567,10 @@ namespace AZ
         {
             AZ_PROFILE_FUNCTION(AzCore);
             AssetManagerNotificationBus::Broadcast(&AssetManagerNotificationBus::Events::OnAssetEventsDispatchBegin);
-            AssetBus::ExecuteQueuedEvents();
+            while (AssetBus::QueuedEventCount())
+            {
+                AssetBus::ExecuteQueuedEvents();
+            }
```

Comment:
This seems potentially risky to get stuck in the `while` forever, or for a long time if events keeps being queued. Why isn't enough with executing the queue events? Maybe some protection could be added to exit the look or at least warn the user if it's stuck here?

Reply:
Well the issue is that the AssetBus was relying on additional events being added during processing, to be pumped at that time.
This `DispatchEvents` function could invoke the `AssetContainer::OnAssetReady` function, which would then broadcast the `OnAssetContainerReady` using the `AssetManagerBus`, result in it being queued to the AssetBus.

It is hard to break of the loop, when it isn't known how many of the queue functions need to execute for the current thread to completely finish loading assets.
Using a timeout would result in a non-stable situation where based on how much is being done, the <num of calls> to DispatchEvents might result.

The issue that I was noticing was that the DispatchEvents loop in [WaitForAsset::Wait](https://github.com/o3de/o3de/blob/5889b6acdd439189348a11a52c1fb94df3aad39f/Code/Framework/AzCore/AzCore/Asset/AssetManager.cpp#L324-L330) was not being called enough times, resulting in a situation at the end of the AssetCatalogTest and AssetManager test left over functions being on the queue.

One thing we could do is to call DispatchEvents one more time at the end of WaitForAsset::Wait after the load has completed, to make sure any left over events are queued.

Currently with the implementation in this PR, it replicates the behavior of `EBusQueuePolicy::Execute` function continuously executing functions on the queue, as long as it is not empty

Reply:
Fair enough, thanks for the explanation. I'll give the approval to move on, but I'm also going to make another suggestion too: 

Since there is still the risk of an infinite loop here from an undesired side effect (for example, a bug that processing event A adds event B, processing event B ultimately adds event A). To mitigate that we could have a timer, not to break the loop, but to print a line in the log (example "Asset bus: still processing events...") every X seconds (whatever we think it's appropriate: every 5secs or 10 secs...)

This way, if there is a bug it will be obvious that it's stuck there, making it easier to spot errors.

Reply:
I was able to remove the while loop in the `AssetManager::DispatchEvents` function 

The `WaitForAsset::Wait` function now has a final `DispatchEvents` call after load has completedto make sure that the `OnAssetReady` callback for the loaded asset has been invoked, which resolves all but 1 of the test failures .

The final fix was to update the `AssetJobsFloodTest.AssetWithNoLoadReference_LoadDependencies_BehaviorObeyed` to invoke the `DispatchEvents` function until it receives an OnAssetContainerReady callback which indicates that an Asset and it's dependencies are completely loaded which fixes that test.